File: prototype/rails/src/app/domain/auth/email_address.rb
module Domain
  module Auth
    # Addresses arrive from sign-in forms and from magic-link rows; both sides
    # compare the normalized form so one person never ends up with two accounts.
    module EmailAddress
      SHAPE = /\A[^@\s]+@[^@\s]+\.[^@\s]+\z/

      module_function

      def normalize(email)
        email.to_s.strip.downcase
      end

      def valid?(email)
        SHAPE.match?(normalize(email))
      end
    end
  end
end

File: prototype/php/src/app/Domain/Auth/EmailAddress.php
<?php

declare(strict_types=1);

namespace App\Domain\Auth;

final class EmailAddress
{
    /**
     * Addresses arrive from forms and from magic-link rows; both sides compare
     * against the normalized form so one person never ends up with two accounts.
     */
    public static function normalize(string $email): string
    {
        return mb_strtolower(trim($email));
    }
}
